File: src/android/ansight-location/src/main/kotlin/ai/ansight/location/AnsightLocationRecorder.kt
package ai.ansight.location

import ai.ansight.runtime.AnsightRuntime
import ai.ansight.runtime.OperationResult
import android.location.Location
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sqrt

class AnsightLocationRecorder @JvmOverloads constructor(
    options: AnsightLocationOptions = AnsightLocationOptions(),
) {
    private val lock = Any()
    private val options = options.normalized()
    private var lastEmittedSample: AnsightLocationSample? = null

    fun record(sample: AnsightLocationSample): OperationResult {
        if (!options.enabled) {
            return OperationResult.failure("Observed location capture is disabled.")
        }
        if (!sample.latitude.isFinite() || sample.latitude !in -90.0..90.0 ||
            !sample.longitude.isFinite() || sample.longitude !in -180.0..180.0
        ) {
            return OperationResult.failure("Observed location coordinates are invalid.")
        }

        val normalized = normalize(sample)
        synchronized(lock) {
            if (shouldSuppress(normalized)) {
                return OperationResult.success("Observed location suppressed by sampling controls.")
            }
            lastEmittedSample = normalized
        }

        val payload = JSONObject()
            .put("schema", Schema)
            .put("sampleId", normalized.sampleId)
            .put("capturedAtUtc", timestamp(normalized.capturedAtEpochMilliseconds))
            .put("source", "app_observed")
            .put("latitude", normalized.latitude)
            .put("longitude", normalized.longitude)
        putOptional(payload, "altitudeMeters", normalized.altitudeMeters)
        putOptional(payload, "horizontalAccuracyMeters", normalized.horizontalAccuracyMeters)
        putOptional(payload, "verticalAccuracyMeters", normalized.verticalAccuracyMeters)
        putOptional(payload, "speedMetersPerSecond", normalized.speedMetersPerSecond)
        putOptional(payload, "headingDegrees", normalized.headingDegrees)
        putOptional(payload, "correlationId", normalized.correlationId)
        putOptional(payload, "runId", normalized.runId)
        return AnsightRuntime.sendSessionEvent(EventType, payload)
    }

    @JvmOverloads
    fun record(
        location: Location,
        correlationId: String? = null,
        runId: String? = null,
    ): OperationResult = record(AnsightLocationSample(
        latitude = location.latitude,
        longitude = location.longitude,
        altitudeMeters = location.altitude.takeIf { location.hasAltitude() },
        horizontalAccuracyMeters = location.accuracy.toDouble().takeIf { location.hasAccuracy() },
        verticalAccuracyMeters = if (android.os.Build.VERSION.SDK_INT >= 26 && location.hasVerticalAccuracy()) {
            location.verticalAccuracyMeters.toDouble()
        } else {
            null
        },
        speedMetersPerSecond = location.speed.toDouble().takeIf { location.hasSpeed() },
        headingDegrees = location.bearing.toDouble().takeIf { location.hasBearing() },
        capturedAtEpochMilliseconds = location.time,
        correlationId = correlationId,
        runId = runId,
    ))

    private fun normalize(sample: AnsightLocationSample): AnsightLocationSample = sample.copy(
        latitude = rounded(sample.latitude),
        longitude = rounded(sample.longitude),
        altitudeMeters = finite(sample.altitudeMeters),
        horizontalAccuracyMeters = nonNegative(sample.horizontalAccuracyMeters),
        verticalAccuracyMeters = nonNegative(sample.verticalAccuracyMeters),
        speedMetersPerSecond = nonNegative(sample.speedMetersPerSecond),
        headingDegrees = finite(sample.headingDegrees),
        sampleId = sample.sampleId.trim().ifEmpty { java.util.UUID.randomUUID().toString() },
        correlationId = sample.correlationId?.trim()?.ifEmpty { null },
        runId = sample.runId?.trim()?.ifEmpty { null },
    )

    private fun shouldSuppress(sample: AnsightLocationSample): Boolean {
        val previous = lastEmittedSample ?: return false
        return sample.capturedAtEpochMilliseconds - previous.capturedAtEpochMilliseconds < options.minimumIntervalMilliseconds ||
            distanceMeters(previous, sample) < options.minimumDistanceMeters
    }

    private fun rounded(value: Double): Double {
        val scale = 10.0.pow(options.decimalPlaces)
        return round(value * scale) / scale
    }

    private fun distanceMeters(first: AnsightLocationSample, second: AnsightLocationSample): Double {
        val radius = 6_371_000.0
        val latitudeDelta = Math.toRadians(second.latitude - first.latitude)
        val longitudeDelta = Math.toRadians(second.longitude - first.longitude)
        val firstLatitude = Math.toRadians(first.latitude)
        val secondLatitude = Math.toRadians(second.latitude)
        val haversine = sin(latitudeDelta / 2).pow(2) +
            cos(firstLatitude) * cos(secondLatitude) * sin(longitudeDelta / 2).pow(2)
        return 2 * radius * asin(sqrt(haversine))
    }

    private fun timestamp(epochMilliseconds: Long): String = SimpleDateFormat(
        "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
        Locale.US,
    ).apply { timeZone = TimeZone.getTimeZone("UTC") }.format(Date(epochMilliseconds))

    private fun finite(value: Double?): Double? = value?.takeIf { it.isFinite() }

    private fun nonNegative(value: Double?): Double? = value?.takeIf { it.isFinite() && it >= 0 }

    private fun putOptional(payload: JSONObject, key: String, value: Any?) {
        if (value != null) payload.put(key, value)
    }

    companion object {
        const val EventType = "CLIENT_LOCATION"
        const val Schema = "ansight.location.sample.v1"
    }
}

File: src/ios/Sources/AnsightLocation/AnsightLocationRecorder.swift
import AnsightCore
import Foundation

public actor AnsightLocationRecorder {
    public static let eventType = "CLIENT_LOCATION"
    public static let schema = "ansight.location.sample.v1"

    private let runtime: AnsightRuntime
    private let options: AnsightLocationOptions
    private var lastEmittedSample: AnsightLocationSample?

    public init(
        runtime: AnsightRuntime = .shared,
        options: AnsightLocationOptions = .init()
    ) {
        self.runtime = runtime
        self.options = options
    }

    public func record(_ sample: AnsightLocationSample) async -> OperationResult {
        guard options.enabled else {
            return .failure("Observed location capture is disabled.")
        }
        guard sample.latitude.isFinite, (-90 ... 90).contains(sample.latitude),
              sample.longitude.isFinite, (-180 ... 180).contains(sample.longitude) else {
            return .failure("Observed location coordinates are invalid.")
        }

        let normalized = normalize(sample)
        if shouldSuppress(normalized) {
            return .success("Observed location suppressed by sampling controls.")
        }
        lastEmittedSample = normalized

        return await runtime.sendSessionEvent(type: Self.eventType, payload: payload(normalized))
    }

    private func normalize(_ sample: AnsightLocationSample) -> AnsightLocationSample {
        AnsightLocationSample(
            latitude: rounded(sample.latitude),
            longitude: rounded(sample.longitude),
            altitudeMeters: finite(sample.altitudeMeters),
            horizontalAccuracyMeters: nonNegative(sample.horizontalAccuracyMeters),
            verticalAccuracyMeters: nonNegative(sample.verticalAccuracyMeters),
            speedMetersPerSecond: nonNegative(sample.speedMetersPerSecond),
            headingDegrees: finite(sample.headingDegrees),
            capturedAt: sample.capturedAt,
            sampleId: normalized(sample.sampleId) ?? UUID().uuidString.lowercased(),
            correlationId: normalized(sample.correlationId),
            runId: normalized(sample.runId)
        )
    }

    private func shouldSuppress(_ sample: AnsightLocationSample) -> Bool {
        guard let previous = lastEmittedSample else { return false }
        return sample.capturedAt.timeIntervalSince(previous.capturedAt) < options.minimumInterval
            || distanceMeters(from: previous, to: sample) < options.minimumDistanceMeters
    }

    private func payload(_ sample: AnsightLocationSample) -> [String: JSONValue] {
        var value: [String: JSONValue] = [
            "schema": .string(Self.schema),
            "sampleId": .string(sample.sampleId ?? UUID().uuidString.lowercased()),
            "capturedAtUtc": .string(Self.timestamp(sample.capturedAt)),
            "source": .string("app_observed"),
            "latitude": .number(sample.latitude),
            "longitude": .number(sample.longitude),
        ]
        add(sample.altitudeMeters, as: "altitudeMeters", to: &value)
        add(sample.horizontalAccuracyMeters, as: "horizontalAccuracyMeters", to: &value)
        add(sample.verticalAccuracyMeters, as: "verticalAccuracyMeters", to: &value)
        add(sample.speedMetersPerSecond, as: "speedMetersPerSecond", to: &value)
        add(sample.headingDegrees, as: "headingDegrees", to: &value)
        add(sample.correlationId, as: "correlationId", to: &value)
        add(sample.runId, as: "runId", to: &value)
        return value
    }

    private func rounded(_ value: Double) -> Double {
        let scale = pow(10, Double(options.decimalPlaces))
        return (value * scale).rounded() / scale
    }

    private func distanceMeters(from: AnsightLocationSample, to: AnsightLocationSample) -> Double {
        let radius = 6_371_000.0
        let latitudeDelta = (to.latitude - from.latitude) * .pi / 180
        let longitudeDelta = (to.longitude - from.longitude) * .pi / 180
        let firstLatitude = from.latitude * .pi / 180
        let secondLatitude = to.latitude * .pi / 180
        let haversine = pow(sin(latitudeDelta / 2), 2)
            + cos(firstLatitude) * cos(secondLatitude) * pow(sin(longitudeDelta / 2), 2)
        return 2 * radius * asin(sqrt(haversine))
    }

    private func finite(_ value: Double?) -> Double? {
        value?.isFinite == true ? value : nil
    }

    private func nonNegative(_ value: Double?) -> Double? {
        value?.isFinite == true && value! >= 0 ? value : nil
    }

    private func normalized(_ value: String?) -> String? {
        let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed?.isEmpty == false ? trimmed : nil
    }

    private func add(_ source: Double?, as key: String, to payload: inout [String: JSONValue]) {
        if let source { payload[key] = .number(source) }
    }

    private func add(_ source: String?, as key: String, to payload: inout [String: JSONValue]) {
        if let source { payload[key] = .string(source) }
    }

    private static func timestamp(_ date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: date)
    }
}
